app/domain/invoices/sac_memberships/extend_roles_for_invoicing.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+#  Copyright (c) 2024, Schweizer Alpen-Club. This file is part of
+#  hitobito_sac_cas and licensed under the Affero General Public License version 3
+#  or later. See the COPYING file at the top-level directory or at
+#  https://github.com/hitobito/hitobito_sac_cas.
+
+module Invoices::SacMemberships
+  class ExtendRolesForInvoicing
+    ROLES_TO_EXTEND = SacCas::MITGLIED_ROLES + [
+      Group::Ehrenmitglieder::Ehrenmitglied,
+      Group::SektionsMitglieder::Beguenstigt,
+      Group::SektionsMitglieder::Ehrenmitglied
+    ].map(&:sti_name)
+
+    def initialize(date)
+      @date = date
+    end
+
+    def extend_roles
+      role_ids = Role.where(type: ROLES_TO_EXTEND, terminated: false, delete_on: ..@date, person_id: person_ids).pluck(:id)
+
+      Role.where(id: role_ids).update_all(delete_on: @date) if role_ids.present?
```

Comment:
While `update_all` is very efficient it skips rails validations. Is this OK @codez ?

Reply:
Das könnte in der Tat ein Problem sein oder, falls künftig Anpassungen an den Callbacks gemacht werden, eines werden. Daher würde ich die Rollen über `find_each` laden und dann regulär Updaten. Das gibt zwar pro Update ein Query, aber immerhin nicht noch je eines beim Laden der Rolle.

Reply:
@codez wenn ich das nutze:
```rb
def extend_roles
  Role.where(type: ROLES_TO_EXTEND, terminated: false, delete_on: ...@date, person_id: person_ids)
    .find_each { |role| role.update!(delete_on: @date) }
end
```

gibt es auch queries für personen und gruppen:
<img width="1031" alt="image" src="https://github.com/user-attachments/assets/95959233-0db0-4bae-a552-795509340a5c">

also der `it "only makes 2 database queries"` spec ist nachher `it "makes 33 database queries"`. ist das nicht etwas zu viel? sonst könnte man ja die erste rolle einzeln updaten und wenn die valid ist die anderen mit `update_all`?

Reply:
Die kannst du wohl mit einem passenden `includes` beim Laden der Roles reduzieren. Ev. reicht `includes(:person, :group)`, je nachdem welche Personen und Gruppen da geladen werden.

Reply:
`includes` geht leider nicht. anscheinend werden bei den validations personen und groups geladen. mit `.find_each { |role| role.update_column(:delete_on, @date) }` passiert dies nämlich nicht, weil es so nicht validated wird

Reply:
Ok, da wird wild alles mögliche validiert. Möglicherweise könnten die Validierungen hier sogar dazwischen funken, wenn die Reihenfolge zufällig ist. Dann machen wir doch `update_all` und lassen die Validierungen sein. 

Damit aber nicht 150'000 Ids geladen werden, würde ich mit Batches arbeiten:

```ruby
Role.where(type: ROLES_TO_EXTEND, terminated: false, delete_on: ...@date, person_id: person_ids).in_batches do |batch|
        Role.where(id: batch.pluck(:id)).update_all(delete_on: @date)
end
```